### PR TITLE
Add Density_prior_box_op

### PR DESCRIPTION
**Pull request by shippingwang**
Density_prior_box is prepared for generating detection box.

### CODE REVIEW

File: paddle/fluid/operators/detection/density_prior_box_op.cc [hunk 1]
```diff
@@ -0,0 +1,228 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (heavengate) [inline paddle/fluid/operators/detection/density_prior_box_op.cc]: 2016->2018

File: paddle/fluid/operators/detection/density_prior_box_op.cc [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+  }
+};
+
+class DensityPriorBoxOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput(
+        "Input",
+        "(Tensor, default Tensor<float>), "
+        "the input feature data of DensityPriorBoxOp, The layout is NCHW.");
```
Comment (heavengate) [inline paddle/fluid/operators/detection/density_prior_box_op.cc]: Second 'The' should be lower case 'the'

File: paddle/fluid/operators/detection/density_prior_box_op.cc [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        "Please note, this order affects the weights order of convolution layer"
+        "followed by and does not affect the final detection results.")
+        .SetDefault(false);
+    AddComment(R"DOC(
+Density Prior box operator
+Generate density prior boxes for SSD(Single Shot MultiBox Detector) algorithm.
+Each position of the input produce N prior boxes, N is determined by
+ the count of min_sizes, max_sizes and aspect_ratios, The size of the
+ box is in range(min_size, max_size) interval, which is generated in
+ sequence according to the aspect_ratios.
```
Comment (heavengate) [inline paddle/fluid/operators/detection/density_prior_box_op.cc]: try to unify the indention?

Comment (heavengate) [reply paddle/fluid/operators/detection/density_prior_box_op.cc]: Maybe you should explain the algorithm more clearly, such as use math formula.

File: paddle/fluid/operators/detection/density_prior_box_op.h [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+      step_height = static_cast<T>(img_height) / feature_height;
+    } else {
+      step_width = step_w;
+      step_height = step_h;
+    }
+
+    int num_priors = aspect_ratios.size() * min_sizes.size();
+
+    if (fixed_sizes.size() > 0) {
+      if (densities.size() > 0) {
```
Comment (heavengate) [inline paddle/fluid/operators/detection/density_prior_box_op.h]: if (fixed_sizes.size() > 0 && densities.size() > 0) will be more clearly.

File: paddle/fluid/operators/detection/density_prior_box_op.h [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+                      (center_y_temp + box_height_ratio / 2.) / img_height <= 1
+                          ? (center_y_temp + box_height_ratio / 2.) / img_height
+                          : 1;
+                  idx++;
+                }
+              }
+            }
+          } else {
+            // if does not have fixed ratios, apply the aspect ratios to
+            // generate density prior boxes instead.
```
Comment (heavengate) [inline paddle/fluid/operators/detection/density_prior_box_op.h]: try to explain that 'step_average' and 'fixed_size' are used in two branches?

**Review by heavengate — COMMENTED** [review]
good work

File: paddle/fluid/operators/detection/density_prior_box_op.cc (hunk 1 above)
Comment (shippingwang) [reply paddle/fluid/operators/detection/density_prior_box_op.cc]: got it

File: paddle/fluid/operators/detection/density_prior_box_op.cc (hunk 2 above)
Comment (shippingwang) [reply paddle/fluid/operators/detection/density_prior_box_op.cc]: got it

File: paddle/fluid/operators/detection/density_prior_box_op.cc (hunk 3 above)
Comment (shippingwang) [reply paddle/fluid/operators/detection/density_prior_box_op.cc]: yeah

File: paddle/fluid/operators/detection/density_prior_box_op.h (hunk 4 above)
Comment (shippingwang) [reply paddle/fluid/operators/detection/density_prior_box_op.h]: copy that

File: python/paddle/fluid/layers/detection.py [hunk 6]
```diff
@@ -1023,6 +1024,164 @@ def _is_list_or_tuple_(data):
...
+                      steps=[0.0, 0.0],
+                      offset=0.5,
+                      name=None,
+                      min_max_aspect_ratios_order=False):
+    """
+    **Density Prior Box Operator**
+
+    Generate density prior boxes for SSD(Single Shot MultiBox Detector) 
+    algorithm. Each position of the input produce N prior boxes, N is 
+    determined by the count of min_sizes, max_sizes, aspect_ratios,
```
Comment (panyx0718) [inline python/paddle/fluid/layers/detection.py]: would be great to give the math. How is N calculated by these arguments.

File: python/paddle/fluid/layers/detection.py [hunk 7]
```diff
@@ -1023,6 +1024,164 @@ def _is_list_or_tuple_(data):
...
+    **Density Prior Box Operator**
+
+    Generate density prior boxes for SSD(Single Shot MultiBox Detector) 
+    algorithm. Each position of the input produce N prior boxes, N is 
+    determined by the count of min_sizes, max_sizes, aspect_ratios, 
+    densities, fixed_sizes and fixed_ratios. Besides boxes generated by
+    prior_box operator, boxes center at grid points arround each input 
+    position is also generated by this operator, and the grid points is
+    determined by densities and the count of density prior box is determined 
+    by fixed_sizes and fixed_ratios.
```
Comment (panyx0718) [inline python/paddle/fluid/layers/detection.py]: Can you use some examples to show how are they "determined"?

File: python/paddle/fluid/layers/detection.py [hunk 8]
```diff
@@ -1023,6 +1024,164 @@ def _is_list_or_tuple_(data):
...
+        The layout is [H, W, num_priors, 4].
+        H is the height of input, W is the width of input
+        num_priors is the total
+        box count of each position of input
+
+
+    Examples:
+        .. code-block:: python
+
+            box, var = fluid.layers.prior_box(
```
Comment (panyx0718) [inline python/paddle/fluid/layers/detection.py]: density_prior_box?

Give some examples?

File: python/paddle/fluid/layers/detection.py (hunk 6 above)
Comment (panyx0718) [reply python/paddle/fluid/layers/detection.py]: If there is a paper, also good to put a link to the paper.

File: python/paddle/fluid/layers/detection.py [hunk 9]
```diff
@@ -1023,6 +1024,164 @@ def _is_list_or_tuple_(data):
...
+                      densities=None,
+                      fixed_sizes=None,
+                      fixed_ratios=None,
+                      variance=[0.1, 0.1, 0.2, 0.2],
+                      flip=False,
+                      clip=False,
+                      steps=[0.0, 0.0],
+                      offset=0.5,
+                      name=None,
+                      min_max_aspect_ratios_order=False):
```
Comment (panyx0718) [inline python/paddle/fluid/layers/detection.py]: what's the difference from prior_box? Is it prior_box with a few more boxes decided by fixed_sizes and density?

File: python/paddle/fluid/layers/detection.py (hunk 7 above)
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: 这里的文档觉得只用和prior_box差异就行了，不用这么复杂。

File: paddle/fluid/operators/detection/density_prior_box_op.h [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+#include <algorithm>
+#include <vector>
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/operators/math/math_function.h"
+#include "paddle/fluid/platform/transform.h"
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/density_prior_box_op.h]: 这个.h里的内容觉得都可以移动到.cc里吧。

File: paddle/fluid/operators/detection/density_prior_box_op.h [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+
+namespace paddle {
+namespace operators {
+
+inline void ExpandAspectRatios(const std::vector<float>& input_aspect_ratior,
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/density_prior_box_op.h]:
```
static inline void 
```

另外，这个和prior_box里的一样吗？可以公用吗？

File: paddle/fluid/operators/detection/density_prior_box_op.h [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+      output_aspect_ratior->push_back(ar);
+      if (flip) {
+        output_aspect_ratior->push_back(1.0f / ar);
+      }
+    }
+  }
+}
+
+template <typename T>
+struct ClipFunctor {
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/density_prior_box_op.h]: 这个可以include proir_box.h里的吧，不用重复一边。

File: paddle/fluid/operators/detection/density_prior_box_op.cc [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+        the count of fixed_sizes, fixed_ratios, densities, min_sizes, max_sizes and aspect_ratios,
+        Generate the density prior boxes by fixed_sizes, fixed rstios, densities, 
+        if the fixed_sizes, fixed_ratios and densities is supported, the calculation of shift is as follows:
+        $$shift = \frac{StepAverage}{density}$$, which step_average is a harmonic value of step_w and step_h
+        if the fixed_ratios is not supported, the calculation of shift is changed to:
+        $$shift= \frac{FixedSize}{density}$$
+        and then generate prior box.
+        )DOC");
+  }
+};
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/density_prior_box_op.cc:164]: 这里其实可以继承PriorBox的ProtoMaker吧

File: python/paddle/fluid/tests/unittests/test_density_prior_box_op.py [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+    def init_test_input(self):
+        self.image = np.random.random(
+            (self.batch_size, self.image_channels, self.image_w,
+             self.image_h)).astype('float32')
+
+        self.input = np.random.random(
+            (self.batch_size, self.input_channels, self.layer_w,
+             self.layer_h)).astype('float32')
+
+    def init_test_output(self):
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_density_prior_box_op.py:92]: 这个核心的逻辑计算，可以提取到`class TestDensityPriorBoxOp(OpTest):`外面，作为一个独立的函数吧。

File: paddle/fluid/operators/detection/density_prior_box_op.cc (hunk 13 above)
Comment (shippingwang) [reply paddle/fluid/operators/detection/density_prior_box_op.cc:164]: > 这里其实可以继承PriorBox的ProtoMaker吧
是的，我改一下

File: paddle/fluid/operators/detection/density_prior_box_op.h (hunk 11 above)
Comment (shippingwang) [reply paddle/fluid/operators/detection/density_prior_box_op.h]: > ```
> static inline void 
> ```
> 另外，这个和prior_box里的一样吗？可以公用吗？

是一样的

File: python/paddle/fluid/layers/detection.py (hunk 9 above)
Comment (shippingwang) [reply python/paddle/fluid/layers/detection.py]: > what's the difference from prior_box? Is it prior_box with a few more boxes decided by fixed_sizes and density?

Right.

File: python/paddle/fluid/layers/detection.py (hunk 6 above)
Comment (shippingwang) [reply python/paddle/fluid/layers/detection.py]: > would be great to give the math. How is N calculated by these arguments.

Yes, I will add an expression

File: python/paddle/fluid/layers/detection.py (hunk 7 above)
Comment (shippingwang) [reply python/paddle/fluid/layers/detection.py]: en

Comment (shippingwang) [reply python/paddle/fluid/layers/detection.py]: sure

File: python/paddle/fluid/tests/unittests/test_density_prior_box_op.py (hunk 14 above)
Comment (shippingwang) [reply python/paddle/fluid/tests/unittests/test_density_prior_box_op.py:92]: > 这个核心的逻辑计算，可以提取到`class TestDensityPriorBoxOp(OpTest):`外面，作为一个独立的函数吧。

好的

File: paddle/fluid/operators/detection/density_prior_box_op.h (hunk 5 above)
Comment (shippingwang) [reply paddle/fluid/operators/detection/density_prior_box_op.h]: > try to explain that 'step_average' and 'fixed_size' are used in two branches?

I talked to the colleague of visual team, and he said they always input the fixed_ratios, so the second branches(apply the fixed_size to compute shift) won't run, but I will still keep it for future modification.

**Review by qingqing01 — APPROVED** [review]

**Review by panyx0718 — APPROVED** [review]
Approve for API. Need more docs

**Pull request merged**